gensim/matutils.py

```diff
@@ -897,10 +897,9 @@ def hellinger(vec1, vec2):
     if isbow(vec1) and isbow(vec2):
         # if it is a BoW format, instead of converting to dense we use dictionaries to calculate appropriate distance
         vec1, vec2 = dict(vec1), dict(vec2)
-        if len(vec2) < len(vec1):
-            vec1, vec2 = vec2, vec1  # swap references so that we iterate over the shorter vector
+        indexs = set(vec1.keys() + vec2.keys())
         sim = np.sqrt(
-            0.5 * sum((np.sqrt(value) - np.sqrt(vec2.get(index, 0.0)))**2 for index, value in iteritems(vec1))
+            0.5 * sum((np.sqrt(vec1.get(index, 0.0)) - np.sqrt(vec2.get(index, 0.0)))**2 for index in indexs)
```

Comment:
Also please add a test for it (you can use an example from original issue).

Reply:
Should I add a test in this function [https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/test_similarity_metrics.py#L106](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/test_similarity_metrics.py#L106) or anywhere else ？ And it seems that the current two different length BOW inputs test in [Line109](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/test_similarity_metrics.py#L109) has a wrong expected value, I'm tring to correct it.

Reply:
Yes, please fix current test & add needed tests to `TestHellinger` class